rust/cadeau/src/xmf/vpx/mod.rs

```diff
@@ -139,57 +139,60 @@ pub struct VpxFrame {
 
 impl Drop for VpxFrame {
     fn drop(&mut self) {
-        // SAFETY: see SAFETY Note in VpxFrame::new, this is safe to call.
+        // SAFETY: see SAFETY Note in VpxFrame::from_raw, this is safe to call.
         unsafe {
             XmfVpxFrame_Destroy(self.ptr);
         }
     }
 }
 
 impl VpxFrame {
-    /// # SAFETY
+    /// # Safety
+    /// 
     /// The pointer must be valid and must not be null.
+    /// Once it's constructed, the struct VpxFrame owns the pointer. As it is a always a deep copy of the frame data from a packet.
+    /// unless the frame is dropped, the pointer is always valid.
```

Comment:
nitpick: The wording is confusing.
If I understand correctly ,`XmfVpxPacket_GetFrame` is making a deep copy, and only the caller actually knows that we are working with a deep copy of the frame data.
The Safety section should documents that a owned pointer is required.

By the way, [a valid pointer is never null. By definition, a null-pointer is invalid](https://doc.rust-lang.org/std/ptr/index.html#safety). Thus "must be valid and non null" is a redundant wording.

Here is an example of how this could be improved:

```c
/**
 * Retrieves the frame data from the packet.
 * This function will return NULL if the packet is not a frame packet.
 *
 * @param packet Pointer to the packet instance.
 * @return Pointer to the frame data, or NULL if the packet is not a frame packet.
 *
 * @note The returned frame data is deep copied from the packet data and should be freed by the caller.
 */
XMF_EXPORT XmfVpxFrame *XmfVpxPacket_GetFrame(const XmfVpxPacket *packet);
```

```rust
/// Constructs a VpxFrame from a raw pointer.
///
/// # Safety
/// 
/// - The pointer must be valid.
/// - The pointer must be owned. (`VpxFrame` is responsible for freeing the resource.)
/// - The pointed `XmfVpxFrame` must be owning the internal data.
unsafe fn from_raw(ptr: *mut XmfVpxFrame) -> Self { … }

// SAFETY: The VpxFrame type guarantees that `ptr` is a valid pointer.
let frame_ptr = unsafe { XmfVpxPacket_GetFrame(self.ptr) };

if frame_ptr.is_null() {
    None
} else {
    // SAFETY: We verified the pointer is not null.
    // XmfVpxPacket_GetFrame always returns a valid, owned pointer.
    // XmfVpxPacket_GetFrame is performing a deep copy of the XmfVpxPacket data into the frame data, which must be freed separately via XmfVpxFrame_Destroy.
    let frame = unsafe { VpxFrame::from_raw(frame_ptr) };
    Some(frame)
}
```

See how the upper layer is relying on the lower layer to prove that the code is correct.

Reply:
updated